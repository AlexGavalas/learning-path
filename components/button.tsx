import type { FC, ButtonHTMLAttributes } from 'react';

type ButtonVariant = 'danger' | 'wrapper';

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    variant?: ButtonVariant;
}

export const Button: FC<ButtonProps> = ({ children, variant, ...rest }) => {
    let bgClass =
        variant === 'danger'
            ? 'bg-red-500 hover:bg-red-700 dark:bg-red-500 dark:hover:bg-red-700'
            : 'bg-gray-200 hover:bg-gray-300 dark:bg-gray-500 dark:hover:bg-gray-600';

    if (variant === 'wrapper') {
        bgClass = 'bg-transparent hover:bg-gray-300 dark:hover:bg-gray-600';
    }

    const textClass =
        variant === 'danger' ? 'text-white' : 'text-black dark:text-white';

    return (
        <button
            {...rest}
            className={`p-2 flex items-center rounded border-0 ${textClass} ${bgClass}`}
        >
            {children}
        </button>
    );
};
