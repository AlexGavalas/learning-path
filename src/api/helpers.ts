import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
