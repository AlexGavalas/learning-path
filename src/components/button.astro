---
type ButtonVariant = 'wrapper' | 'default';

const bgClassNamesPerVariant: Record<ButtonVariant, string> = {
    default:
        'bg-slate-300 hover:bg-slate-300/75 dark:bg-slate-600 dark:hover:bg-slate-600/75',
    wrapper: 'bg-transparent hover:bg-slate-300 dark:hover:bg-slate-600',
};

const { variant, ...rest } = Astro.props;

let bgClass = bgClassNamesPerVariant.default;
---

<button {...rest} class={`flex items-center rounded border-0 p-2 ${bgClass}`}>
    <slot />
</button>
