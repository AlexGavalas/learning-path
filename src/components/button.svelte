<script lang="ts">
    import type { HTMLButtonAttributes } from 'svelte/elements';

    type ButtonVariant = 'wrapper' | 'default';

    interface $$Props extends HTMLButtonAttributes {
        variant?: ButtonVariant;
        extraClass?: string;
    }

    const bgClassNamesPerVariant: Record<ButtonVariant, string> = {
        default:
            'bg-slate-300 hover:bg-slate-300/75 dark:bg-slate-600 dark:hover:bg-slate-600/75',
        wrapper: 'bg-transparent hover:bg-slate-300 dark:hover:bg-slate-600',
    };

    export let variant: ButtonVariant = 'default';
    export let extraClass = '';

    let bgClass = bgClassNamesPerVariant[variant];
</script>

<button
    {...$$restProps}
    on:click
    class={`flex items-center rounded border-0 p-2 ${bgClass} ${extraClass}`}
>
    <slot />
</button>
