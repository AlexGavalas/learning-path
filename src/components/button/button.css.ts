import { style, styleVariants } from '@vanilla-extract/css';

import { vars } from '../../styles/theme.css';

const base = style({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: vars.spacing.xs,
    border: 'none',
    color: vars.color.background,
    padding: vars.spacing.sm,
    ':active': {
        opacity: 0.8,
    },
});

const defaultVariant = style({
    backgroundColor: `oklch(from ${vars.color.primary} l c h)`,
    ':hover': {
        backgroundColor: `oklch(from ${vars.color.primary} calc(l + 0.1) c h)`,
    },
});

const wrapperVariant = style({
    backgroundColor: 'transparent',
    ':hover': {
        backgroundColor: `oklch(from ${vars.color.background} calc(l + 0.07) c h)`,
    },
});

export const button = styleVariants({
    default: [base, defaultVariant],
    wrapper: [base, wrapperVariant],
});
