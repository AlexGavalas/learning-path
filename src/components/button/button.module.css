.base {
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: var(--spacing-xs);
    border: none;
    color: var(--color-background);
    padding: var(--spacing-sm);

    &:active {
        opacity: 0.8;
    }

    &.default {
        background-color: var(--color-primary);

        &:hover {
            background-color: oklch(
                from var(--color-primary) calc(l + 0.1) c h
            );
        }
    }

    &.wrapper {
        background-color: transparent;

        &:hover {
            background-color: oklch(
                from var(--color-background) calc(l + 0.07) c h
            );
        }
    }
}
