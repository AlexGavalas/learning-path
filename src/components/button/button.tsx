import type { ButtonHTMLAttributes, FC } from 'react';

type ButtonVariant = 'wrapper' | 'default';

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
    variant?: ButtonVariant;
};

const bgClassNamesPerVariant: Record<ButtonVariant, string> = {
    default:
        'bg-slate-300 hover:bg-slate-300/75 dark:bg-slate-600 dark:hover:bg-slate-600/75',
    wrapper: 'bg-transparent hover:bg-slate-300 dark:hover:bg-slate-600',
};

export const Button: FC<ButtonProps> = ({
    children,
    variant = 'default',
    ...rest
}) => {
    const bgClass = bgClassNamesPerVariant[variant];

    return (
        <button
            {...rest}
            className={`flex items-center rounded border-0 p-2 ${bgClass}`}
        >
            {children}
        </button>
    );
};
