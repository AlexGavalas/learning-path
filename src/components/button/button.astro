---
import type { HTMLAttributes } from 'astro/types';
import * as cn from './button.css';

type ButtonVariant = 'wrapper' | 'default';

interface Props extends HTMLAttributes<'button'> {
    variant?: ButtonVariant;
}

const { variant = 'default', class: extraClass, ...rest } = Astro.props;
---

<button {...rest} class:list={[cn.button[variant], extraClass]}>
    <slot />
</button>
