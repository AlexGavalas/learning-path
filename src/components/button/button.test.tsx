import { axe } from 'jest-axe';

import { renderWithUser } from '~test/helpers';

import { Button } from './button';

describe('<Button />', () => {
    it('renders', () => {
        const { container } = renderWithUser(<Button>click me</Button>);

        expect(container).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const { container } = renderWithUser(<Button>click me</Button>);

        const a11yResults = await axe(container);

        expect(a11yResults).toHaveNoViolations();
    });
});
