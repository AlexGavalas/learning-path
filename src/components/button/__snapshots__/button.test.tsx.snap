// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders 1`] = `
<div>
  <button
    class="flex items-center rounded border-0 p-2 bg-slate-300 hover:bg-slate-300/75 dark:bg-slate-600 dark:hover:bg-slate-600/75"
  >
    click me
  </button>
</div>
`;
